File: solutions/go/collatz-conjecture/1/collatz_conjecture.go
package collatzconjecture

import "errors"

func collatzConjectureStep(n int, steps int) int {
	if n == 1 {
        return steps
    } else if n % 2 == 0 {
		return collatzConjectureStep(n/2, steps+1)  
    } else {
        return collatzConjectureStep(3*n +1, steps+1)  
    }
}

func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
        return 0, errors.New("Not a positive number")
    }
    result := collatzConjectureStep(n, 0)
    return result, nil

}

File: solutions/elixir/collatz-conjecture/1/lib/collatz_conjecture.ex
defmodule Integer.Guards do
  defguard is_even(value) when is_integer(value) and rem(value, 2) == 0
  defguard is_odd(value) when is_integer(value) and rem(value, 2) != 0
  defguard is_positive_integer(value) when is_integer(value) and value > 0 
end

defmodule CollatzConjecture do
  @moduledoc """
  Calculate steps of Collatz Conjecture
  """

  import Integer.Guards

  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(input) when is_positive_integer(input) do
    step(input, 0)
  end
  
  defp step(1, steps) do
    steps
  end
  
  defp step(input, steps) when is_even(input) do
    step(div(input, 2), steps + 1)
  end

  defp step(input, steps) when is_odd(input) do
    step((3 * input ) + 1, steps + 1)
  end
  
end
